Iter13

itaraxa opened the pull request.

itaraxa committed:
doc: Added new tasks

itaraxa committed:
feat: added usage of context

itaraxa committed:
feat: added database support with connection verification

itaraxa committed:
test: added context.TODO into test functions

itaraxa committed:
fix: corrected issues reported by the go vet

itaraxa committed:
fix(server): corrected import mistake

itaraxa committed:
feat(server): added postgres-db storage realisation

itaraxa committed:
chore: update version

itaraxa committed:
feat(server): updated use cases to work with the database using context

itaraxa committed:
feat(logger): added Fatal logger method

itaraxa committed:
chore(config): added environment variables to the help output

itaraxa committed:
refactor(handlers): added using of constants and contexts into handlers

itaraxa committed:
feat(memstorage): added Close() method that clears memory

itaraxa committed:
refactor(server): replaced os.Exit() calls with return

itaraxa committed:
chore: updated .gitignore

itaraxa committed:
feat(postgres): corrected SQL queries to the database

itaraxa committed:
feat(postgres): added GetAll, String and HTML methods

itaraxa committed:
feat(server): disabled file saving when using database storage

itaraxa committed:
fix: resolved golangci-lint issues – removed unused interfaces

itaraxa committed:
fix: resolved go-vet issues – renamed constants

itaraxa committed:
chore: updated version number

itaraxa committed:
feat(server): added ability to receive metrics in batches

itaraxa committed:
feat(agent): added ability to send metrics in batches

itaraxa committed:
fix(server): corrected the returned error code

itaraxa committed:
fix(server): corrected processing new counters

itaraxa committed:
debug(postgres): added printing all metrics info into stdout

itaraxa committed:
debug(storage): added printing metrics from restoring file into stdout

itaraxa committed:
debug(storage): fix printing counter

itaraxa committed:
feat(server): added storage cleanup before restoring saved metrics from the file

itaraxa committed:
fix(memstorage): corected using of mutex in Clear() method

itaraxa committed:
fix(postgres): added mutexes to control concurrent access

itaraxa committed:
feat: added logging of batch update request

itaraxa committed:
feat(server): added ability to handle duplicate metrics in a single batch

itaraxa committed:
chore: added description for increment 13 and updated version

itaraxa committed:
feat(agent): added retry capability in the agent application for failed sends

itaraxa committed:
feat(server): added retry capability for database requests in the server application

itaraxa committed:
fix(server): corrected pgErr codes for retry

itaraxa committed:
feat(server): added HTTP server shutdown on Ctrl+C signal
feat(server): added context passing from main to handlers and methods
docs: added comments
chore: removed debug information output

bulbetski commented on cmd/agent/main.go:
```diff
@@ -78,22 +79,23 @@ func main() {
-		panic(err)
+		fmt.Println(err)
+		return
 	}
```
В main функции можно сразу завершать программу через log.Fatal. Использование метода fmt.Println не принято в продакш-реди приложениях

bulbetski commented on cmd/server/main.go:
```diff
@@ -66,42 +68,59 @@ func (sa *ServerApp) Run() {
+		if err := services.SaveMetricsToFile(ctx, sa.logger, sa.storage, sa.config.FileStoragePath); err == nil {
 			sa.logger.Info("metric data has been saved to the file", "filename", sa.config.FileStoragePath)
 		} else {
 			sa.logger.Error("metric data hasn't been saved to the file", "error", err.Error(), "filename", sa.config.FileStoragePath)
```
Два лога в одном err != nil условии. Инфо надо вынести отсюда

bulbetski commented on cmd/server/main.go:
```diff
@@ -66,42 +68,59 @@ func (sa *ServerApp) Run() {
+	// если воостанавливаем метрики из файла, то предварительно очищаем хранилище
 	if sa.config.Restore {
+		sa.logger.Info("clear storage")
+		ctx3s, cancel3s := context.WithTimeout(ctx, 3*time.Second)
```
можно просто назвать ctxWithTimeout

bulbetski commented on cmd/server/main.go:
```diff
@@ -110,53 +129,76 @@ func (sa *ServerApp) Run() {
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "error parsing environment variable: %v", err)
-		os.Exit(1)
+		return
```
Стоит завершать через log.Fatal

bulbetski commented on cmd/server/main.go:
```diff
@@ -165,10 +207,19 @@ func main() {
+	} else {
+		s := memstorage.NewMemStorage()
+		app := NewServerApp(logger, s, r, serverConf)
+		app.Run()
```
Это должно происходить вне зависимости от выбора хранилища. То есть от наличия дсна зависит только то, какой репозиторий мы используем

bulbetski commented on internal/config/agent.go:
```diff
@@ -28,11 +29,13 @@ func NewAgentConfig() *AgentConfig {
 		LogLevel:       `INFO`,
 		ReportMode:     `json`,
 		Compress:       `gzip`,
+		Batch:          true,
```
мб это как-то конфигурировать?

itaraxa replied:
В ТЗ таких требований не было, я ввел флаг для проверки работы агента в старых режимах - кроме как размера батча ничего в голову не приходит

bulbetski commented on internal/handlers/dbhandlers.go:
```diff
@@ -0,0 +1,45 @@
+	return func(w http.ResponseWriter, req *http.Request) {
+		if req.Method != http.MethodGet {
+			http.Error(w, "wrong request type != GET", http.StatusMethodNotAllowed)
+			return
```
Проверка метода HTTP-запроса выглядит не очень гибко. Если в будущем появится необходимость поддерживать другие методы, потребуется доработка. Сам маршрутизатор по сути обрабатывает запрос и если в маршрутизаторе метод записан в гет, то только гет и будет.

bulbetski commented on internal/handlers/dbhandlers.go:
```diff
@@ -0,0 +1,45 @@
+			return
+		}
+		l.Info("received a request to ping db-storage")
+		w.Header().Set("Content-Type", "text/html")
```
Мы вообще ничего не пишем в ответ, но тем не менее хэдер какой-то ставим. Либо не нужен хэдер, либо нужно записать ответ

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -58,7 +81,7 @@ func GetAllCurrentMetrics(s metricPrinter, l logger) http.HandlerFunc {
-		_, err := w.Write([]byte(s.HTML()))
+		_, err := w.Write([]byte(s.HTML(ctx)))
 		if err != nil {
 			l.Error("cannot write HTML to response body")
```
Понимаю, что в этом месте по сути функция завершается, но стоит сделать аналогично первой ошибке. запись хттп ошибку в ответ и сделать return

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -70,18 +93,19 @@ GetMetrica creates a handler that returns the metric value
 	http.HandlerFunc
 */
-func GetMetrica(s metricGetter, l logger) http.HandlerFunc {
+func GetMetrica(ctx context.Context, s metricGetter, l logger) http.HandlerFunc {
```
Логирование происходит в начале функции, но значения из `chi.URLParam` извлекаются многократно. Можго вынесенти параметры в переменные перед логированием, чтобы избежать повторного извлечения.
		metricType := chi.URLParam(req, "type")
		metricName := chi.URLParam(req, "name")

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -91,29 +115,33 @@ func GetMetrica(s metricGetter, l logger) http.HandlerFunc {
 			res = fmt.Sprint(v.(float64))
-		case "counter":
+		case counter:
 			res = fmt.Sprint(v.(int64))
```
Преобразования типов могут приводить к панике, надо проверять на ok еще.

Также можно вообще не создавать переменную res, а сразу писать в ответ w.Write(что-то там)

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -136,24 +164,24 @@ func JSONGetMetrica(s metricGetter, l logger) http.HandlerFunc {
-		case "counter":
+		case counter:
 			t := valueFromStorage.(int64)
 			jm.Delta = &t
```
преобразование типов может привести к панике

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -334,3 +373,61 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+		}
+		// Processing
+		var buf bytes.Buffer
+		_, err := buf.ReadFrom(req.Body)
```
если нет тела запроса, это вызовет панику, надо првоерить на нил

itaraxa replied:
Добавил проверку

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -334,3 +373,61 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+			return
+		}
+
+		w.Header().Set("Content-Type", "applicttion/json")
```
опечатка

itaraxa replied:
исправил

bulbetski commented on internal/repositories/memstorage/memstorage.go:
```diff
@@ -35,6 +36,16 @@ func (m *MemStorage) UpdateGauge(metricName string, value float64) error {
+	MetricValue *float64
+}) error {
+	for _, metric := range metrics {
+		m.Gauge[metric.MetricName] = *metric.MetricValue
```
MetricValue может быть nil, что вызовет панику

itaraxa replied:
Добавил проверку и переход к следующей итерации цикла

bulbetski commented on internal/repositories/memstorage/memstorage.go:
```diff
@@ -47,12 +58,26 @@ Returns:
+}) error {
+	for _, metric := range metrics {
+		if currentDelta, ok := m.Counter[metric.MetricName]; ok {
+			m.Counter[metric.MetricName] = currentDelta + *metric.MetricDelta
+		} else {
+			m.Counter[metric.MetricName] = *metric.MetricDelta
```
MetricDelta может быть nil, что вызовет панику

bulbetski commented on internal/repositories/memstorage/memstorage.go:
```diff
@@ -224,3 +249,19 @@ func NewMemStorage() *MemStorage {
 }
+
+func (m *MemStorage) PingContext(ctx context.Context) error {
+	return nil
```
Можно проверить, инициализирована ли мапа, но можно этого и не делать

bulbetski commented on internal/repositories/memstorage/memstorage_test.go:
```diff
@@ -68,7 +69,7 @@ func TestMemStorage_UpdateGauge(t *testing.T) {
 				Counter: tt.fields.Counter,
 			}
-			if err := m.UpdateGauge(tt.args.metricName, tt.args.value); (err != nil) != tt.wantErr {
+			if err := m.UpdateGauge(context.TODO(), tt.args.metricName, tt.args.value); (err != nil) != tt.wantErr {
```
использование context.TODO() не лучшая практика. Стоит понять, какой контекст тут лучше использовать

bulbetski commented on internal/repositories/postgres/migration.go:
```diff
@@ -0,0 +1,40 @@
+func prepareTablesContext(ctx context.Context, db *sql.DB) error {
+	files, _ := migrationFiles.ReadDir("migrations")
+	for _, file := range files {
+		content, _ := migrationFiles.ReadFile("migrations/" + file.Name())
```
не обрабатываем ошибку

itaraxa replied:
Добавил проверку возвращаемой ошибки

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+func NewPostgresRepository(ctx context.Context, databaseURL string) (*PostgresRepository, error) {
+	db, err := sql.Open("pgx", databaseURL)
+	if err != nil {
+		return &PostgresRepository{}, err
```
тут надо возвращать nil ошибку

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+
+	err = prepareTablesContext(ctx5s, db)
+	if err != nil {
+		return &PostgresRepository{}, fmt.Errorf("cannot create tables in database storage: %w", err)
```
аналогично, тут надо возвращать nil ошибку

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+
+	_, err := pr.db.ExecContext(ctx, "INSERT INTO gauges (metric_id, metric_value) VALUES ($1, $2);", metricName, value)
+	if err != nil {
+		fmt.Println(err.Error())
```
вместо fmt.Println лучше обычное логирование

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+		} else {
+			err = tx.Commit()
+		}
+	}()
```
этот код повторяется, можно вынести в отдельный метод

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+	err = tx.QueryRowContext(ctx, "SELECT metric_delta FROM counters WHERE metric_id = $1 ORDER BY metric_timestamp DESC LIMIT 1;", metricName).Scan(&currentDelta)
+	if err != nil {
+		// delta-value not in db
+		if err == sql.ErrNoRows {
```
Лучше проверять через errors.Is

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+		err = tx.QueryRowContext(ctx, "SELECT metric_delta FROM counters WHERE metric_id = $1 ORDER BY metric_timestamp DESC LIMIT 1;", metric.MetricName).Scan(&currentDelta)
+		if err != nil {
+			// delta-value not in db
+			if err == sql.ErrNoRows {
```
лучше проверять через errors.Is

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+	s := ""
+	metrics, err := pr.GetAllMetrics(ctx)
+	if err != nil {
+		fmt.Printf("error getting metrics from postgres storage: %v", err.Error())
```
лучше логировать через logger

bulbetski commented on internal/repositories/postgres/postgres.go:
```diff
@@ -0,0 +1,518 @@
+
+	metrics, err := pr.GetAllMetrics(ctx)
+	if err != nil {
+		fmt.Printf("error getting metrics from postgres storage: %v", err.Error())
```
лучше логировать через logger

bulbetski commented on internal/services/agent.go:
```diff
@@ -91,6 +103,34 @@ func sendMetricaToServerJSON(l logger, ms MetricsGetter, serverURL string, clien
 }
 
+func sendMetricaToServerBatch(l logger, ms MetricsGetter, serverURL string, client *http.Client) error {
+	jsonDataReq, err := json.Marshal(ms.GetData())
```
Нет проверки на пустой `ms.GetData()`. Если метрики отсутствуют, создается пустой JSON-запрос. Это может привести к ошибкам на сервере.

bulbetski commented on internal/services/agent.go:
```diff
@@ -91,6 +103,34 @@ func sendMetricaToServerJSON(l logger, ms MetricsGetter, serverURL string, clien
+		return err
+	}
+	l.Debug("json data for send", "string representation", string(jsonDataReq))
+	req, err := http.NewRequest(`POST`, fmt.Sprintf("http://%s/updates/", serverURL), bytes.NewBuffer(jsonDataReq))
```
Если `serverURL` уже содержит протокол, это может привести к двойному `http://`. Так что лучше это указать где-нибудь при иниицализации

itaraxa replied:
Думаю вынести эту проверку в функцию валидации пакета config

bulbetski replied:
достаточно быть увереным, что оно где-то установлено. Желательно при инициализации клиента. В месте вызова уже об этом думать не надо. 
Либо сделать хелпер для создания запросов и в нем уже делать валидацию

itaraxa replied:
добавил функцию createUrl

bulbetski commented on internal/services/agent.go:
```diff
@@ -165,6 +205,113 @@ func sendMetricaToServerJSONgzip(l logger, ms MetricsGetter, serverURL string, c
 }
 
+func sendMetricaToServerBatchgzip(l logger, ms MetricsGetter, serverURL string, client *http.Client) error {
+	jsonDataReq, err := json.Marshal(ms.GetData())
```
Проверка на пустой ms.GetData()

bulbetski commented on internal/services/agent.go:
```diff
@@ -165,6 +205,113 @@ func sendMetricaToServerJSONgzip(l logger, ms MetricsGetter, serverURL string, c
+	}
+	jsonGzipDataReq, err := compress(jsonDataReq)
+	if err != nil {
+		l.Error("cannot compress data", "error", err.Error())
```
нет return'a

bulbetski commented on internal/services/agent.go:
```diff
@@ -165,6 +205,113 @@ func sendMetricaToServerJSONgzip(l logger, ms MetricsGetter, serverURL string, c
+	*http.Response
+	error
+*/
+func retryDo(l logger, client *http.Client, req *http.Request) (*http.Response, error) {
```
Код явно повторяет один и тот же блок с небольшими изменениями (номер попытки и таймаут). Это ухудшает читаемость и поддержку. Стоит подумать, как избавиться от дублирования

itaraxa replied:
Реализовал по аналогии с ретраем запроса к базе данных - через функцию-обертку

bulbetski commented on internal/services/server.go:
```diff
@@ -143,17 +160,18 @@ WriteMetrics saves metrics from storage and current timestamp to the any writer
 	blob := make(map[string]interface{})
 	blob["timestamp"] = time.Now()
-	blob["metrics"] = mg.GetAllMetrics()
+	blob["metrics"], _ = mg.GetAllMetrics(ctx)
```
Не обрабатываем ошибку

bulbetski commented on internal/services/utility.go:
```diff
@@ -3,15 +3,21 @@ package services
 
 func ShowStorage(s MetricPrinter) {
-	fmt.Println(s.String())
+	fmt.Println(s.String(context.TODO()))
```
Точно ли тут нужны принты через fmt

itaraxa replied:
Эти функции использовались для отладки, переделал их на возврат строки

bulbetski reviewed (changes requested):
В целом код нормальный, но есть пара мест, которые требуют доработок, по типу приведения типов и обработки ошибок

itaraxa committed:
refactor(server): In case of an error in the main() function, log.Fatal() is used

itaraxa committed:
refactor(agent): eplaced fmt.Print() and return with log.Fatal. Added error handling

itaraxa committed:
refactor: rename ctx var

itaraxa committed:
refactor: removed useless checks in handlers

itaraxa committed:
refactor(handlers): added type asserting errors checks and logging

itaraxa committed:
refactor(memstorage): added check for nil metric value

itaraxa committed:
refactor(postgres): improved error handling

itaraxa committed:
refactor(agent): added check for empty metric slice

itaraxa committed:
refactor(server): added error handling

itaraxa committed:
fix: fixxed go-vet issue

bulbetski commented on cmd/agent/main.go:
```diff
@@ -77,23 +79,21 @@ func main() {
 	}
 	if agentConf.ShowVersion {
 		fmt.Println(version.AgentVersion)
-		os.Exit(0)
+		return
```
пропустил log.Fatal

itaraxa replied:
Я не считаю что в этом случае нужен log.Fatal. Пользователь указал ключ для того, что бы приложение только вывело версию. Приложение выводит версию и завершает работу

bulbetski replied:
да, ты прав, я фигню написал

bulbetski commented on cmd/server/main.go:
```diff
@@ -66,42 +68,62 @@ func (sa *ServerApp) Run() {
+	go func(cancel context.CancelFunc) {
+		defer cancel()
+		defer func() {
+			_ = sa.storage.Close()
```
Можно залогировать ошибку. Кароч всегда какая-то обработка должна быть

itaraxa replied:
Добавил логирование

bulbetski commented on cmd/server/main.go:
```diff
@@ -110,65 +132,95 @@ func (sa *ServerApp) Run() {
 	if serverConf.ShowVersion {
-		fmt.Println(version.ServerVersion)
-		os.Exit(0)
+		log.Print(version.ServerVersion)
+		return
```
log.Fatal

itaraxa replied:
Аналогично с таим же блоком в прложении агента. Считаю что это штатное поведение, и оно не должно возвращать ошибку

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -311,26 +349,94 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+			if g, ok := value.(float64); ok {
+				resp.Value = &g
+			} else {
+				l.Error("type assertion -> float64", "value", value)
```
не надо ли нам завершить выполнение функции в этом случае?

itaraxa replied:
Надо, добавил return

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -311,26 +349,94 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+				l.Error("type assertion -> float64", "value", value)
+			}
+		case counter:
 			c, _ := value.(int64)
 			resp.Delta = &c
```
а зачем это делать для каунтера? ну и опять же может быть паника

itaraxa replied:
Забыл удалить

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -311,26 +349,94 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+			if c, ok := value.(int64); ok {
+				resp.Delta = &c
+			} else {
+				l.Error("type assertion -> int64", "value", value)
```
не надо ли нам завершить выполнение функции в этом случае?

itaraxa replied:
Надо, добавли return

bulbetski commented on internal/handlers/handlers.go:
```diff
@@ -311,26 +349,94 @@ func JSONUpdateHandler(l logger, s metricStorager) http.HandlerFunc {
+			}
 		}
 
 		body, _ := json.Marshal(resp)
```
пропущена обработка ошибки

itaraxa replied:
добавлена обработка ошибки

bulbetski commented on internal/services/server_test.go:
```diff
@@ -80,121 +79,121 @@ func TestParseQueryString(t *testing.T) {
-		})
-	}
-}
+// func TestUpdateMetrica(t *testing.T) {
```
Оно нам надо, раз оно закоменчено?

bulbetski reviewed (commented):
Ну, большинство комментов не исправлено. Надо поправлять, особенно важный с функцией retryDo

itaraxa committed:
feat: added function for sending retry requests
feat: added function for forming request URLs
refactor: extracted transaction initialization, commit, or rollback into a separate function
feat: added error checking and logging

bulbetski reviewed (approved):
Всё супер исправил. Код хороший
